Mathlib/Topology/CWComplex.lean

```diff
@@ -0,0 +1,164 @@
+union of the n-spheres. -/
+def sigmaAttachMap (X : TopCat) (n : ℤ) (cells : Type) (attach_maps : cells → C(𝕊 n, X)) :
+    TopCat.of (Σ (_ : cells), 𝕊 n) ⟶ X where
+  toFun := fun ⟨i, x⟩ ↦ attach_maps i x
+  continuous_toFun := continuous_sigma fun i ↦ (attach_maps i).continuous_toFun
+
+/-- A type witnessing that X' is obtained from X by attaching (n+1)-disks -/
+structure AttachCells (X X' : TopCat) (n : ℤ) where
+  /-- The index type over the (n+1)-disks -/
+  cells : Type
+  /-- For each (n+1)-disk, we have an attaching map from its boundary, namely an n-sphere,
+  to `X`. -/
+  attach_maps : cells → C(𝕊 n, X)
+  /-- X' is the pushout obtained from X along `sigmaAttachMap`. -/
+  iso_pushout : X' ≅ Limits.pushout
+    (sigmaSphereInclusion n cells) (sigmaAttachMap X n cells attach_maps)
+
+end CWComplex
+
+/-- A relative CW-complex contains an expanding sequence of subspaces `sk i`
+(called the `i`-skeleta) for `i ≥ -1`, where `sk (-1)` is an arbitrary topological space,
+isomorphic to `A`, and each `sk (n+1)` is obtained from `sk n` by attaching (n+1)-disks. -/
+structure RelativeCWComplex (A : TopCat) where
+  /-- Skeleta -/
+  sk : ℤ → TopCat
+  /-- A is isomorphic to the (-1)-skeleton. -/
+  iso_sk_neg_one : A ≅ sk (-1)
+  /-- The (n+1)-skeleton is obtained from the n-skeleton by attaching (n+1)-disks. -/
+  attach_cells : (n : ℤ) → CWComplex.AttachCells (sk n) (sk (n + 1)) n
+
+/-- A CW-complex is a relative CW-complex whose (-1)-skeleton is empty. -/
+abbrev CWComplex := RelativeCWComplex (TopCat.of Empty)
+
+namespace CWComplex
+
+noncomputable section Topology
+
+/-- The inclusion map from `X` to `X'`, given that `X'` is obtained from X by attaching
+(n+1)-disks -/
+def attachCellsInclusion (X X' : TopCat) (n : ℤ) (att : AttachCells X X' n) : X ⟶ X' :=
+  @Limits.pushout.inr TopCat _ _ _ X
+    (sigmaSphereInclusion n att.cells)
+    (sigmaAttachMap X n att.cells att.attach_maps) _ ≫ att.iso_pushout.inv
+
+/-- The inclusion map from the n-skeleton to the (n+1)-skeleton of a relative CW-complex -/
+def skeletaInclusion {A : TopCat} (X : RelativeCWComplex A) (n : ℤ) : X.sk n ⟶ X.sk (n + 1) :=
+  attachCellsInclusion (X.sk n) (X.sk (n + 1)) n (X.attach_cells n)
+
+/-- The inclusion map from the n-skeleton to the m-skeleton of a relative CW-complex -/
+def skeletaInclusion' {A : TopCat} (X : RelativeCWComplex A)
+    (n : ℤ) (m : ℤ) (n_le_m : n ≤ m) : X.sk n ⟶ X.sk m :=
+  if h : n = m then by
+    subst m
+    exact 𝟙 (X.sk n)
+  else by
+    have h' : n < m := Int.lt_iff_le_and_ne.mpr ⟨n_le_m, h⟩
+    exact skeletaInclusion X n ≫ skeletaInclusion' X (n + 1) m h'
+  termination_by Int.toNat (m - n)
+  decreasing_by
+    simp_wf
+    rw [Int.toNat_of_nonneg (Int.sub_nonneg_of_le h')]
+    linarith
+
+/-- The colimit diagram in the definition of a relative CW-complex -/
+def colimitDiagram {A : TopCat} (X : RelativeCWComplex A) : ℤ ⥤ TopCat where
+  obj := X.sk
+  map := @fun n m n_le_m => skeletaInclusion' X n m <| Quiver.Hom.le n_le_m
```

Comment:
As a general remark, you may favour the following syntax:
```suggestion
  map n m n_le_m := skeletaInclusion' X n m <| Quiver.Hom.le n_le_m
```
Instead of `Quiver.Hom.le`, I think you could use `leOfHom`.

Reply:
Thanks. Since I have updated the definition of the topology on a relative CW-complex using your `Functor.OfSequence`, this `colimitDiagram` is no longer needed.